lib/mongodb.nit

```diff
@@ -0,0 +1,708 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# MongoDB Nit Driver.
+#
+# This is actually a wrapper around the [MongoDB C Driver](http://api.mongodb.org/c/1.1.4/index.html).
+#
+# Usage:
+#
+# ~~~
+# # Opens the connexion with the Mongo server.
+# var client = new MongoClient("mongodb://localhost:27017/")
+#
+# # Retrieves a collection.
+# var col = client.database("test").collection("test")
+#
+# # Inserts a document in the collection.
+# var doc = new JsonObject
+# doc["foo"] = 10
+# doc["bar"] = "bar"
+# doc["baz"] = new JsonArray
+# assert col.insert(doc)
+#
+# # Retrieves a document from the collection.
+# var query = new JsonObject
+# query["foo"] = 10
+# var res = col.find(query)
+# assert doc["bar"] == "bar"
+#
+# # Closes the connexion
+# client.close
+# ~~~
+#
+# TODO
+#
+# * Get last Object_ID
+# * Handle cursors
+module mongodb is pkgconfig("libmongoc-1.0")
+
+import json
+
+in "C header" `{
+	#include <mongoc.h>
+`}
+
+# Eveything inside MongoDB is manipulated as BSON Objects.
+#
+# Wraps [bson_t](http://api.mongodb.org/libbson/1.1.4/bson_t.html)
+#
+# See:
+# * [Binary JSON spec](http://bsonspec.org/)
+# * [Libbson](http://api.mongodb.org/libbson/1.1.4/)
+extern class BSON `{ bson_t * `}
+
+	# Returns a new empty BSON object.
+	#
+	# ~~~
+	# var bson = new BSON
+	# assert bson.to_s == """{ }"""
+	# ~~~
+	new `{ return bson_new(); `}
+
+	# Returns a new BSON object initialized from the content of `json`.
+	#
+	# ~~~
+	# var obj = new JsonObject
+	# obj["age"] = 10
+	# obj["name"] = "Rick"
+	# obj["ELS"] = new JsonArray
+	# var bson = new BSON.from_json(obj)
+	# assert bson.to_s == """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# ~~~
+	new from_json(json: JsonObject) do
+		return new BSON.from_json_string(json.to_json)
+	end
+
+	# Returns a new BSON object parsed from `json_string`.
+	#
+	# If `json_string` is not a valid JSON string, this initializer returns NULL.
+	#
+	# ~~~
+	# var str = """{ "age" : 10, "name" : "Rick", "ELS" : [  ] }"""
+	# var bson = new BSON.from_json_string(str)
+	# assert bson.to_s == str
+	# ~~~
+	new from_json_string(json_string: String) import String.to_cstring `{
+		bson_error_t error;
+		bson_t *bson;
+		bson = bson_new_from_json(String_to_cstring(json_string), -1, &error);
+		if(!bson) {
+			MongoDb_set_mongoc_error(recv, &error);
+	        return NULL;
+		}
+		return bson;
+	`}
+
+	redef fun to_s import NativeString.to_s `{
+		char *str = bson_as_json(recv, NULL);
+		return NativeString_to_s(str);
```

Comment:
There's a potential memory leak here. There is nothing that really helps in the mongodb API. But you can do something like:

~~~
char *str = bson_as_json(recv, NULL);
String nit_str = NativeString_to_s_with_copy(str)
bson_free(str);
return nit_str;
~~~